### DISCUSSION TITLE
Chat with PDF, Doc

### DISCUSSION THREAD

**Discussion by aiaicode** [ggml-org/llama.cpp#4021] [Q&A]
One of the biggest use case of LLMs especially for businesses is chatting with PDF and Docs privately. 

Would it be difficult to add this as feature in llama.cpp?

**Comment by staviq** [comment]
> Would it be difficult

Long story short, yes.

Most documents would overflow the context just by themselves, so "traditional" approaches are out.

Typically, this is done with vector storage which is way out of scope of llama.cpp.

Check out Private GPT: https://github.com/imartinez/privateGPT

**Reply by Galunid** [reply]
There's also https://github.com/PromtEngineer/localGPT/ (it uses llama-cpp-python)

**Comment by matthiasdg** [comment]
What you're looking for is RAG (retrieval-augmented generation). The PrivateGPT mentioned before has examples, there is also https://python.langchain.com/docs/modules/data_connection/
Though it's not necessary to use langchain of course. Process is often a bit like this: extract text from your documents, cut it up in smaller parts, vectorize these using an embeddings model and store these in a vector db. Then also vectorize your prompt and get the closest vectors/pieces of text (different distance measures available) from the db. Then put these as context in your prompt

**Reply by ossirytk** [reply]
I've written a character AI chatbot in python as an example for this exact thing. Using langchain, chainlit, llama-cpp-python and chroma. My example uses text files as documents, but langchain has a pdf document parser. https://github.com/ossirytk/llama-cpp-chat-memory

**Reply by x4080** [reply]
Hi @ossirytk, whats your opinion about RAG in chat situation. For example We have chat about :
```
Who is Biden 's son
```
and then using RAG, it will answer
```
his son name is Hunter
```
then we ask again
```
How old is he ?
```

The question now is what's the sentence to search in RAG database, since now the context which is 'Hunter' is separated from the question : 'How old is he ?'

I have this problem with context, do you have solution for it ?

TIA

**Reply by ossirytk** [reply]
I'm currently thinking about problems like these. I find that the basic vanilla RAG is pretty limited, and I'm thinking on how to make it better. What I have so far is to use NER to parse metadata for more accurate searches and some thoughts on linking summaries to more expansive documents. I haven't really thought about your scenario. Maybe matching metadata to the current conversation history of maybe three last messages.

**Reply by PuxAI** [reply]
I was actually thinking about the same issue/situation. To RAG or Not to RAG :) 
Perhaps its as easy like this:
create an agent that is able to run mutliple agents, one for the base model (chat/instruct), one for RAG and one for websearch.
If the base model doesnt know, follow up with a vectordb query, if the vectordb doesnt have any similarity to the query, perform a websearch (duckduckgo)

**Reply by x4080** [reply]
Thanks for answering, yeah its pretty tricky to use chat with documents

**Reply by ossirytk** [reply]
@PuxAI The main problem I have is that it can be difficult to know when the information is not in the vector database. Since normally I use complete sentences to ask questions, and there is almost always some kind of familiarity in the sentence structure. Common words and sentences inevitably cause static in the familiarity search and it can be difficult to filter out. I'm still trying to figure out how to make reliable searches. Or how to construct good vector databases.

**Reply by staviq** [reply]
The same way you modify the conversation in the background to let the model use data from vector database, ask the model to re-phrase user's question in form of a Google search query first, use that on the vector database, and then, clean up the conversation as usuall.

You don't know if a sentence will match vectordb without context, but the model does, because it can extrapolate missing information in form of a full sentence question.

Make the model itself do the heavy lifting :)

As in, the model itself should be able to reconstruct you question "how old is he" into "how old is Bidet's son, Hunter?", which should yield much better vectordb matches.

**Reply by ossirytk** [reply]
This is basically the solution that I've come to. The other solution I have is to use AI to parse the vectordb documents for metadata automatically. With good metadata linking the document subjects, the search accuracy increases accordingly.

**Reply by x4080** [reply]
@staviq Do you train the model to convert phrases from current and prev conversation to "keywords" that can rephrase the question ?
